### DISCUSSION TITLE
How to mock the AppServiceProvider app->isProduction()?

### DISCUSSION THREAD

**Discussion by hfoletto** [pestphp/pest#1156] [Q&A]
Hello there, greetings!

I'm increasing the safety of my application by using the `Model::shouldBeStrict()` method in my AppServiceProvider.
That's working great, however specifically for the `preventLazyLoading` part, I'd like to handle that automatically in production, so no errors are thrown for this specific problem. This is what I have:

```php
// Everything strict, all the time.
Model::shouldBeStrict();

// In production, merely log lazy loading violations.
if ($this->app->isProduction()) {
    Model::handleLazyLoadingViolationUsing(function (Model $model, string $relation) {
        $class = get_class($model);

        info("Attempted to lazy load [{$relation}] on model [{$class}].");
    });
}
```

I added the following test to make sure this is working:
```php
test('N+1 queries are not allowed', function () {
    User::factory()->count(2)->create();
    $output = [];
    $users = User::all();

    foreach ($users as $user) {
        // `address` is lazy loaded.
        $output[] = $user->name.' - '.$user->address->street_address_line_1;
    }
})->throws(LazyLoadingViolationException::class);
```

So far so good, however I'm facing problems when trying to assert that it does allow it in production. I have the following test:
```php
test('N+1 queries are allowed in production', function () {
    config(['app.env' => 'production']);

    User::factory()->count(2)->create();
    $output = [];
    $users = User::all();

    foreach ($users as $user) {
        // `address` is lazy loaded.
        $output[] = $user->name.' - '.$user->address->street_address_line_1;
    }
})->throwsNoExceptions();
```

However it doesn't work. It throws the exception anyways.
Just to debug it, I tried to add the following log to the AppServiceProvider:
```php
// Everything strict, all the time.
Model::shouldBeStrict();

logger('Logging inside AppServiceProvider. isProduction = '.($this->app->isProduction() ? 'true' : 'false'));

// In production, merely log lazy loading violations.
if ($this->app->isProduction()) {
    Model::handleLazyLoadingViolationUsing(function (Model $model, string $relation) {
        $class = get_class($model);

        info("Attempted to lazy load [{$relation}] on model [{$class}].");
    });
}
```
It always log that it's `false`.

What is the correct way to test this?

Thank you!

**Comment by ipontt-neering** [comment]
Partially mock the `Config` facade instead. Make it return whatever `isProduction()` is checking for.

The example here uses the `Cache` facade but the syntax shouldn't change.

https://laravel.com/docs/12.x/mocking#mocking-facades
